Coset crate

Cesar199999 opened the pull request:
This PR, developed in collaboration with @Antonio95, implements the `p3-coset` crate which, together with `p3-poly`, provides essential building blocks for implementing the STIR low-degree test in Plonky3. By separating this functionality into a dedicated crate, we aim to make the main PR more modular and easier to review. The history of the `p3-coset` crate can still be tracked in the STIR PR: https://github.com/Plonky3/Plonky3/pull/674

The `README.md` file contains a high-level overview of the crate and its functionality.

## Implementation Details

- The core of this crate is the `TwoAdicCoset` struct, which represents a coset of size 2^n (where n is some non-negative integer) in a `TwoAdicField` F. To create a new instance, you can use the `new` constructor by providing a `shift` and `log_size`. The generator is automatically set to `F::two_adic_generator(log_size)`. 

- The struct maintains two important fields: `dft` and `generator_iter_squares`. The `dft` field is lazily initialized, meaning it's only computed when needed for the first polynomial evaluation or interpolation. The `generator_iter_squares` field serves to cache squares of the generator (see the point below).

- Cosets can be indexed in two ways, using the `element` and `element_immutable` methods. Both methods return the i-th element of the coset (i.e. `shift * generator^i`), the difference is that `element` caches the intermediate squares of the generator (`generator^(2^0), generator^(2^1), ..., generator^(2^n)`) and thus requires a mutable reference to `self`.

- To maintain modularity and simplify the review process, we have designed the codebase to be independent from `p3-poly`. As a result, the `TwoAdicCoset` struct intentionally avoids any direct dependencies on `p3-poly`. This architectural decision means that `evaluate_polynomial` and `interpolate_polynomial` work directly with vectors of polynomial coefficients instead of `Polynomial` objects. While this approach means that `interpolate_polynomial` may return coefficient vectors that aren't in canonical form (containing leading zeros), this isn't a concern since `Polynomial::from_coeffs` handles zero-trimming automatically when converting the vector to a proper polynomial object.

- Cosets are not given in "canonical form", i.e. the same coset can be represented by different shifts. This is a deliberate design choice to allow for more flexibility in the implementation of the STIR low-degree test, in particular when the coset shift is equal to the subgroup generator, which allows for a special case in the evaluation and interpolation algorithms (detailed in the `README.md`).

SyxtonPrime commented:
Hey, thanks for all the work on this and the other PR's. These are excellent.

Regarding `TwoAdicCoset`, I think this already actually exists in the code base. We have a struct: `TwoAdicMultiplicativeCoset` inside commit/src/domain. These two structs should be merged.

Antonio95 commented on coset/src/lib.rs:
```diff
@@ -0,0 +1,467 @@
+    generator: F,
+    shift: F,
+    consumed: bool,
+}
```
Good call, thanks. We've fixed it.

It turns out that the correct structure to use was `Take<Powers<F>` instead of just `Powers<F>`, because `some_iterator.take(n)` does not return `SomeIterator` but `Take<SomeIterator>`. The latter structure cannot be converted into `SomeInterator` (with things like `.iter()`, `into_iter()` or `into()`) because the two are fundamentally incompatible (basically, `Take<SomeIterator>` wraps around `SomeIterator` but enriches it with a counter to keep track of when the cap is reached). This is essentially the same reason we needed the `consumed` flag in the original structure.

Antonio95 commented on coset/src/lib.rs:
```diff
@@ -0,0 +1,467 @@
+        // for each divisor n of |F| - 1, and its elements e are uniquely
+        // caracterised by the condition e^n = 1.
+
+        // We check (shift * element)^(2^log_size) = 1, terminating early if
```
Good catch, we've fixed it

Antonio95 commented on coset/src/lib.rs:
```diff
@@ -0,0 +1,467 @@
+        e == F::ONE
+    }
+
+    /// Returns the coefficients of the unique polynomial of degree less than
+    /// `2^log_size` that interpolates the given evaluations on the coset (with
+    /// the canonical order "shift * g^0, shift * g^1, ...`).
+    ///
+    /// This function receives a mutable reference to `self` because it memoises
+    /// the FFT machinery internally if not previously set.
+    ///
+    /// Note that the vector of coefficients returned by this function might
+    /// contain leading zeros. However, `Polynomial::from_coeffs` (in `p3-poly`)
+    /// will automatically remove them.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the number of evaluations is greater than the size of the
+    /// coset.
+    pub fn interpolate(&mut self, evals: Vec<F>) -> Vec<F> {
+        let size = 1 << self.log_size;
+
+        assert!(
+            evals.len() == size,
+            "The number of evaluations must be equal to the size of the coset."
+        );
+
+        let dft = self.dft.get_or_insert_with(Radix2Dit::default);
+        dft.coset_idft(evals, self.shift)
+    }
```
Are you referring to why this function exists, or to some specific point of its internals? If the question is the former, I guess this abstracts handling the FFT from the user of the library, but this is admittedly a very tenuous reason when "handling" just means: `let dft = Radix2DiT.default();`. Cf. `domain.interpolate(evals)` vs. `dft.coset_idft(evals, domain.shift)` (plus the latter has the advantage that it doesn't force the user into one specific implementor of `TwoAdicSubgroupDft`). We'd be happy to get rid of this method if deem it unnecessary.

SyxtonPrime replied:
Yeah I think my vote would be to store a `dft` in the `STIR` struct instead? (E.g. similarly to how a `dft` is stored in `TwoAdicFriPcs`?.)

I think this should let you get rid of this function.

Cesar199999 commented:
> Hey, thanks for all the work on this and the other PR's. These are excellent.
> 
> Regarding `TwoAdicCoset`, I think this already actually exists in the code base. We have a struct: `TwoAdicMultiplicativeCoset` inside commit/src/domain. These two structs should be merged.

Hi @SyxtonPrime, thanks for catching this! We agree we should avoid duplicating the existing struct. However, perhaps it makes sense to have a dedicated crate for coset-related functionality (p3-coset) rather than having it in commit/src/domain (this location is a bit obscure, as illustrated by the fact that we didn’t even see it before we implemented our own).

We'd be happy to, for instance, merge all available functionality into `TwoAdicMultiplicativeCoset` (either in commit/src/domain or in p3-coset):

* The current impl block of our TwoAdicCoset (which subsumes the  impl block of TwoAdicMultiplicativeCoset consisting only of the `gen()` method).
* The `impl PolynomialSpace` of the preexisting PolynomialSpace.

I personally believe having coset functionality in the `commit` crate isn't ideal from an architectural perspective, as cosets are a more general mathematical concept that could be useful in other contexts. Moving this functionality to a dedicated coset crate would make it more reusable and maintain better separation of concerns. We're happy to put it wherever you think is best, thought.

What are your thoughts on this approach?

Antonio95 commented on coset/src/lib.rs:
```diff
@@ -0,0 +1,467 @@
+    /// interpolation. This is automatically done when evaluation or
+    /// interpolation are first called, but this method is useful for e. g.
+    /// benchmarking purposes.
+    pub fn initialise_fft(&mut self) {
+        self.dft = Some(Radix2Dit::default());
+    }
```
That sounds good, the issue is that this structure wants to store the FFT machinery internally (for efficiency and, especially, in order to make the call to `domain.evaluate_polynomial(poly.coeffs())` more transparent). There are a couple of options:


* Make the structure generic on `Dft: TwoAdicSubgroupDft`. A downside of this is that it would force users who want to use the domain in a much simpler way (e. g. just for querying elements) to deal with a generic that is unrelated to their functionality of interest (as the compiler could not infer the type)
* Have the structure contain an `impl TwoAdicSubgroupDft`, e. g. in some kind of `Box`. This is a bit uglier, but it would avoid the issue from the previous point. Note this ugliness would be hidden from the user.
* Remove the internal `dft` field altogether. This would require users to have to pass a new `dft` argument to `domain.evaluate_polynomial`, which creates a bit more usage friction.
* Leave the structure as it is (including `dft`, evaluate_polynomial and `initialise_fft`) and simple add two new alternative methods `evaluate_polynomial_custom_dft` (and also `interpolate_with_custom_dft` if we decide to leave `interpolate` in) which receives as an argument the DFT to use.

None of this seems ideal, we're happy to go whichever way you prefer. I would incline to something like:

* Have the structure contain an `impl TwoAdicSubgroupDft`, e. g. in some kind of Box.
* `evaluate_polynomial` would not receive any DFT-related arguments. It would use `self.dft` if `Some`, otherwise setting it to `Some(Radix2Dit::default())` and using that one.
* At any point, the user always would the option to call `initialise_fft(&mut self, dft: impl TwoAdicSubgroupDft)` to set which DFT should be used from that point on.

@Cesar199999 and I have put together two examples of how this last option would work. Crucially, both of them work in all three cases: when `// Option 1` is commented out (so that the trait does not play any role), when `// Option 2` is commented out (which uses the trait) and when both are uncommented. There are two possible designs:

* One without lifetimes at the cost of having a Box in the structure and having `initialise_dft` receive a `Box` (one cannot directly receive the trait object as it has unknown size); Playground: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=22481ca052ca08017d378cc1bb071903
* One with lifetimes which avoids `Box`es altogether. Some toying around seems to show these lifetimes are not problematic in this case, but we’ll only be sure once we make the change and test STIR. Playground: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=88773d2fb4e263fbd2501edaac179834

Let us know your thoughts on all this :)

SyxtonPrime replied:
I commented this elsewhere but I think the cleanest solution is to store the `dft` in the `STIR` struct instead? (E.g. similarly to how a `dft` is stored in `TwoAdicFriPcs`?.).

This should let you get rid of `initialise_fft` and either delete `evaluate_polynomial` or move it into the `STIR` crate.

unzvfu commented on coset/src/lib.rs:
```diff
@@ -0,0 +1,467 @@
+    // Letting s = shift, and g = generator (of order 2^log_size), the coset in
+    // question is
+    //     s * <g> = {s, s * g, shift * g^2, ..., s * g^(2^log_size - 1)]
+    generator: F,
```
Note that recently-merged #690 provided precomputed generators for Goldilocks, so there is no longer a performance benefit to storing the generator separately here.

Antonio95 replied:
Thanks for the heads-up, we weren't aware of this recent update. I guess one could look at the situation this way:

- The main pro of keeping the memoisation optimisation from the PIR is independence/future-proofing from concrete implementations of `TwoAdicField`.

- The main cons I can think of are:
  - The memory cost of storing the elements (which the `TwoAdicField` might even have as constants)
  - Requiring `&mut self` in the method `element(index)` (and the arguably inelegantly named `element_immutable`)
  - As a more minor one, the small computational cost to trim the list of memoised values when the coset is shrunk (with `coset.shrink(log_factor)`)

Let us know if you think the cons outweigh the pros and you'd like us to remove the memoisation, in which case we will also delete the related tests and rename `element_immutable` to `element` (and have it request the squares of the generator from the `TwoAdicField` instead of computing them)

SyxtonPrime replied:
I think it should be safe to assume that `F::two_adic_generator` will be implemented in a reasonably efficient way. (The only field currently which has a slow implementation right now is `BN-254` but I don't think that's a problem).

Given that, I think we should use `F::two_adic_generator`.

Cesar199999 commented:
Hi @SyxtonPrime, just wanted to follow up on this PR as it's been about a month without updates. We're hoping to get this and the related polynomial crate PRs wrapped up before shifting focus to the STIR PR potential changes. Let me know if you need any additional information from @Antonio95 or myself to help move this forward. Thanks!

SyxtonPrime commented:
Ahh apologies for taking so long to get back to you.

I think we should add this as a subset of the Field crate. (So you can call it with `p3_field::coset`).

Would you two (@Cesar199999 and @Antonio95) be able to rework this to merge it with `TwoAdicMultiplicativeCoset` and move it into the right place?

Regarding the other two PR's I think with the changes suggested in the STIR PR, we no longer need to add the Polynomial crate would you be happy to close that one without merging it? (While there is some useful functionality in there, a lot of it is going to be hard/impossible to vectorize and potentially a little slow)

I'd definitely keen to merge the STIR PR, I think the main set of changes which need to be made to that would be:
- Rework the algorithm to remove Dependency on Polynomial crate.
- Minor refactor so it works as a drop in replacement for `TwoAdicFriPcs` in our prover.

Does that seem like a reasonably roadmap? If you two can make those changes I'm happy to re-review.

Antonio95 commented:
Thanks for the reply, @SyxtonPrime, we are happy to implement the changes you describe to this PR. There are three points open in the conversation (by which I mean our last messages therein are awaiting your decision):

- https://github.com/Plonky3/Plonky3/pull/676#discussion_r1971687797
- https://github.com/Plonky3/Plonky3/pull/676#discussion_r1963144586
- https://github.com/Plonky3/Plonky3/pull/676#discussion_r1963315760

Let us know how to address these and we should have all changes ready rather quickly :)

SyxtonPrime commented:
Ahh great, thanks for letting me know. I've gone back an addressed each of those. Let me know if any of my answers doesn't make sense.

Antonio95 commented:
@SyxtonPrime Thanks for addressing all the questions. We have now implemented all of the requested changes to this PR. Let us know if this looks good and, if so, we'll port it over to STIR :)

SyxtonPrime commented on field/src/coset.rs:
```diff
@@ -0,0 +1,252 @@
+/// order) or not.
+impl<F: TwoAdicField> PartialEq for TwoAdicMultiplicativeCoset<F> {
+    fn eq(&self, other: &Self) -> bool {
+        if self.generator() == other.generator() && self.shift == other.shift {
```
Given that `self.generator()` now immediately calls `F::two_adic_generator(self.log_size)`, it's simpler (and slightly cheaper) to just check `self.log_size == other.log_size` and `self.contains(other.shift)`.

SyxtonPrime replied:
I think there is an open question here as to if we should consider two cosets whose elements are the same but in a different order as equal.

Also, where is this used? If it's not used I'd lean towards removing this and Eq.

Cesar199999 replied:
Thanks for raising this point. The `PartialEq` and `Eq` implementations were primarily added for testing purposes, with the thought they might be useful in the future. Since they're not effectively used in the STIR implementation, We've removed them and slightly refactored the affected tests and examples.

Antonio95 replied:
I agree it's unclear whether equality should be implemented as that of sets or tuples. Furthermore, this ambiguity could cause problems when users of the library used `==` assuming it does "the natural thing" (which would be whatever they need in their concrete use case).

Equality was only used in a doc example [here](https://github.com/Plonky3/Plonky3/pull/676/files/2eb8a7c680062d3a25b8d5f5cbc3e3e863f295a5#diff-6960445ef0d8c5aa1b77ba563bce8aa29e42e8b9ffc88630be1a2075a5d4f928R37-R38). It is also not used in (the) STIR (PR).

We have removed `Eq` and `PartialEq` and reworked the doc example to compare the shift and generator individually. If the need arises in the future to compare cosets as sets, one could always take the code from this commit.

SyxtonPrime commented on field/src/coset.rs:
```diff
@@ -0,0 +1,252 @@
+// This falls back to PartialEq
+impl<F: TwoAdicField> Eq for TwoAdicMultiplicativeCoset<F> {}
+
+impl<F: TwoAdicField> IntoIterator for TwoAdicMultiplicativeCoset<F> {
+    type Item = F;
+    type IntoIter = Take<Powers<F>>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.generator()
+            .shifted_powers(self.shift)
+            .take(1 << self.log_size)
+    }
+}
```
Is this just a convenience method so you don't need to call `.iter()`? I'm not sure if we need this either but if we do want it we can simplify the code slightly:
```suggestion
impl<F: TwoAdicField> IntoIterator for TwoAdicMultiplicativeCoset<F> {
    type Item = F;
    type IntoIter = Take<Powers<F>>;

    fn into_iter(self) -> Self::IntoIter {
        self.iter()
    }
}
```

Antonio95 replied:
Indeed, it was just a convenience method to be able to write things like `for element in coset {...}`. Thanks for your clean suggestion, we have incorporated it.

SyxtonPrime commented on field/src/coset.rs:
```diff
@@ -0,0 +1,252 @@
+    ///
+    ///  If `F: TwoAdicField` does not provide an element of order `2 ^
+    /// log_size` (e. g. if `2 ^ log_size` does not divide `|F| - 1`)
+    pub fn new(shift: F, log_size: usize) -> Self {
+        assert!(
+            log_size <= F::TWO_ADICITY,
+            "log_size must be <= the two_adicity of the field ({})",
+            F::TWO_ADICITY
+        );
+
+        Self { shift, log_size }
+    }
```
I comment this elsewhere but I think it might be nice for this to return `Option<Self>` instead of panicking.

SyxtonPrime commented on field/src/coset.rs:
```diff
@@ -0,0 +1,252 @@
+        Self { shift, log_size }
+    }
+
+    /// Returns the generator of the coset.
+    #[inline]
+    pub fn generator(&self) -> F {
+        F::two_adic_generator(self.log_size)
+    }
```
Technically this is really the generator of the subgroup of order self.size().

Could we rename this to `subgroup_generator()`?

Cesar199999 replied:
Sure, it's fixed now

Antonio95 replied:
Done! We have also fixed the comment, which was incorrect no matter how one spins things.

SyxtonPrime reviewed (commented):
Hey this is looking good, most of my comments are pretty minor.

The two slightly larger changes are:
- Can you move `PolynomialSpace` into the field crate? (Either in the same file as `TwoAdicMultiplicativeCoset` or a different file but the same module)
- I've recently been trying to cut down on places where Plonky3 can panic (With the exception of places which are key bottlenecks). Do you think you can change some of the `TwoAdicMultiplicativeCoset` code to output `Option<TwoAdicMultiplicativeCoset>`  instead of failing? I've tried to highlight the main functions.

Cesar199999 commented on field/src/coset.rs:
```diff
@@ -0,0 +1,252 @@
+        }
+
+        e == F::ONE
+    }
```
Thanks for the suggestion! We just refactored the method, should be fixed now

Antonio95 replied:
So, just to confirm, one expects 1 inversion + 1 product to be more expensive than somewhere between 1 and `log_size` squarings, correct? At least for the fields supported in this codebase (or rather, their concrete implementations). We did not have any intuition for these costs as we wrote the function.

SyxtonPrime replied:
Yeah, the current way that inversion is coded is `x^{-1} = x^{p - 2}` so roughly speaking the cost of an inverse is something like `1.5*log_p` squares/multiplies (Which have roughly the same cost). The exact value depends on the number of `1`'s in the binary decomposition of `p - 2` and on the optimal addition chains we can find but, at the very least it always involves at least `log(p)` square/multiply operations. We did look into inversion using other methods (e.g. the Euclidean Algorithm) but that turned out to be slower.

For some concrete numbers, you can check out the `try_inverse` functions in the individual field crates:
| Field  | BabyBear | KoalaBear | Mersenne31 | Goldilocks | 
--------|-----------|-------------|--------------|-------------- |
| #ops | 37 | 36 | 37 | 72 |

So generally avoiding inverses wherever possible is worth it. Two notable exceptions to this rule are `.halve()` and `batch_inverse`. `.halve()` has a custom fast implementation and so doesn't involve inverting `2`. `batch_inverse` is a method for inverting a large number of field elements at once using only `1` field inversion (Each extra element you add to invert costs only `~2` extra multiplications).

Antonio95 replied:
That makes a lot of sense, thank you :) We will incorporate halve() into the STIR PR now that we know about it (instead of passing `two_inverse: Option<F>` around).

Cesar199999 commented on field/src/coset.rs:
```diff
@@ -0,0 +1,252 @@
+            Consider passing the equivalent index % (1 << log_size) instead"
+        );
+        self.shift * self.generator_exp(index)
+    }
```
Thanks for the suggestion! Applying the mask at the beginning yields a much cleaner implementation both here and in `generator_exp`.

Antonio95 replied:
This is a great call, thanks. Implemented.

Antonio95 commented on field/src/coset.rs:
```diff
@@ -0,0 +1,252 @@
+    pub fn shrink_coset(&self, log_scale_factor: usize) -> TwoAdicMultiplicativeCoset<F> {
+        let new_coset = self.shrink_subgroup(log_scale_factor);
+        new_coset.set_shift(self.shift.exp_power_of_2(log_scale_factor))
+    }
```
`fold_coset` would in my opinion be semantically too restrictive in the sense that it automatically ties (the semantics of) `TwoAdicCoset` to folding schemes. The case could be made that that's all this structure is meant to be used for (e. g. FRI and the coming STIR), but it could be good to leave `TwoAdicCoset` as a mathematical abstraction with wider applicability.

I do not quite agree about "shrinking" implying that the input is a subset of the input. Think of e. g. shrinking a circumference, or shrinking the interval [4, 8] by a factor of 2.

In any case, I know I'm just splitting hairs here. We've changed the method to `coset.exp_power_of_2(n)`, since all it returns is `coset^(2^n)` (element-wise) - this mirrors the existing nomenclature for field elements. However, if you think this is too verbose or too different from it's sister function `shrink_coset` (formerly `shrink_subgroup`), we'll be happy to give it any name you prefer 👍

SyxtonPrime replied:
Yeah I mostly agree with everything you've written here. I think `exp_power_of_2` is an excellent name for it.

Cesar199999 commented:
@SyxtonPrime Thank you from @Antonio95 and myself for the thorough review. We've implemented most of your suggestions, including replacing panics with `Option`returns. Regarding moving `PolynomialSpace` into the field crate, we're awaiting your take on this point (https://github.com/Plonky3/Plonky3/pull/676#discussion_r2015952324). Once we have clarity on that, we hope to move forward with addressing the remaining comments and concerns on the STIR PR.

Cesar199999 commented:
@SyxtonPrime - We've been trying to fix the CI pipeline for some time now without success. What's particularly frustrating is that everything runs with no problem on our local environments, both @Antonio95's MacOS and my Ubuntu machine. We've made sure we're both using cargo 1.85.1 and have cleared our caches multiple times to rule out environment issues. The error message from the pipeline is a bit strange. It has a problem with `rand::rng` being a module, yet `rand::rng()` is called successfully in numerous places throughout the codebase (such as baby-bear/src/poseidon2.src#478) without triggering similar errors in the pipeline. This inconsistent behavior between identical function calls is what we cannot figure out. Do you have any insights on what might be causing this discrepancy between CI and our local environments? Any guidance would be greatly appreciated. Thanks!

SyxtonPrime commented:
Ahh I think this might be coming from a clash with PR #740 which slightly changed how RNG worked. I think this should be an easy fix, just add the imports:
```
use rand::rngs::SmallRng;
use rand::{Rng, SeedableRng};
```
and then explicitly initialize the rng with 
```
let mut rng = SmallRng::seed_from_u64(1);
```
(Feel free to use a seed other than `1` if you want).

(I suspect the reason its working on your devices it you haven't merged in the recent main changes).

SyxtonPrime commented on commit/src/domain.rs:
```diff
@@ -1,6 +1,7 @@
 use alloc::vec::Vec;
```
Ahh that's annoying but you are definitely right.

I might try and do some sort of refactor later (Maybe make a new `domain` crate which houses `PolynomialSpace, TwoAdicMultiplicativeCoset` and `CircleDomain`?), either way that's definitely out of the scope of this PR so lets leave it as is for now.

Cesar199999 replied:
Thanks! We agree, refactoring those three components into a dedicated crate makes the most sense. Let us know if you'd like any assistance with that refactoring effort after the STIR PR is merged.

SyxtonPrime commented on field/src/coset.rs:
```diff
@@ -0,0 +1,222 @@
+        }
+    }
+
+    /// Returns the generator of the subgroup whose shift is `self`.
```
I'm not quite sure what you mean by: the subgroup whose shift is `self`.
```suggestion
    /// Returns the generator of the subgroup of order `self.size()`.
```

Antonio95 replied:
The thinking behind the original comment was: `self` is a coset of the form `s * <g>`, so it is a shift of the subgroup `<g>` (of the group of units of the field).

At any rate, @Cesar199999 also got confused when he read that original comment so it's clear my wording wasn't the best. Happy to switch to your proposed wording (already committed) 👍

SyxtonPrime commented on field/src/coset.rs:
```diff
@@ -0,0 +1,222 @@
+    /// Returns the element `shift * generator^index`, which is the `index %
+    /// self.size()`-th element of `self` (and, in particular, the `index`-th
+    /// element of `self` whenever `index` < self.size()).
+    #[inline]
+    pub fn element(&mut self, index: usize) -> F {
+        self.shift * self.generator_exp(index & ((1 << self.log_size) - 1))
+    }
```
Given the change to `generator_exp` you don't need the `&` here.
```suggestion
    #[inline]
    pub fn element(&mut self, index: usize) -> F {
        self.shift * self.generator_exp(index)
    }
```

SyxtonPrime commented on field/src/coset.rs:
```diff
@@ -0,0 +1,222 @@
+    // stored), i. e. rather "fetch-and-multiply".
+    fn generator_exp(&mut self, exp: usize) -> F {
+        let mut gen_power = F::ONE;
+        let mut exp = exp & ((1 << self.log_size()) - 1);
```
Up to you but thought this could do with a short comment for clarity. Also we can use `self.size()` instead of `(1 << self.log_size()) - 1`. (I realize that `(1 << self.log_size()) - 1` was my suggestion originally, just noticed the improvement now)
```suggestion
        // As `generator` satisfies `generator^{self.size()} == 1` we can replace `exp` by `exp mod self.size()`.
        // As `self.size()` is a power of `2` this can be done with an `&` instead of a `%`.
        let mut exp = exp & (self.size() - 1);
```

SyxtonPrime commented on field/src/coset.rs:
```diff
@@ -0,0 +1,222 @@
+        gen_power
+    }
+
+    /// Returns an iterator over the elements of the coset in the canonical order
+    /// "shift * generator^0, shift * generator^1, ...,
+    /// shift * generator^(2^log_size - 1)`.
```
typo
```suggestion
    /// Returns an iterator over the elements of the coset in the canonical order
    /// `shift * generator^0, shift * generator^1, ...,
    /// shift * generator^(2^log_size - 1)`.
```

SyxtonPrime commented on fri/src/two_adic_pcs.rs:
```diff
@@ -148,11 +149,7 @@ where
-            log_n,
-            shift: Val::ONE,
-        }
+        TwoAdicMultiplicativeCoset::new(Val::ONE, log2_strict_usize(degree)).unwrap()
```
In a future PR I'll update `natural_domain_for_degree` to output `Option<Self::Domain>` but for now can you add a short comment saying when this `.unwrap()` will panic.

Antonio95 replied:
When adding the comment we noticed the exact same line exists here: https://github.com/Plonky3/Plonky3/blob/31cfc8b8be83c16500ffbf9f65515ec6b2b83041/commit/src/testing.rs#L58

We have added the same comment therein although that line was not pointed out in the suggestions (maybe because that file is just for testing?). Let us know if you would like us to remove this additional comment.

SyxtonPrime replied:
Yeah I was fine with either adding it or not given that it's in the testing crate so it less of a problem if it fails. Your solution is good.

SyxtonPrime commented on commit/src/domain.rs:
```diff
@@ -178,10 +160,7 @@ impl<Val: TwoAdicField> PolynomialSpace for TwoAdicMultiplicativeCoset<Val> {
-            log_n: log2_ceil_usize(min_size),
-            shift: self.shift * Val::GENERATOR,
-        }
+        Self::new(self.shift() * Val::GENERATOR, log2_ceil_usize(min_size)).unwrap()
```
In a future PR I'll update `create_disjoint_domain` to output Option<Self::Domain> but for now can you add a short comment saying when this `.unwrap()` will panic.

SyxtonPrime commented on commit/src/domain.rs:
```diff
@@ -190,11 +169,14 @@ impl<Val: TwoAdicField> PolynomialSpace for TwoAdicMultiplicativeCoset<Val> {
+                    self.shift() * self.subgroup_generator().exp_u64(i as u64),
+                    self.log_size() - log_chunks,
+                )
+                .unwrap()
```
It's pretty obvious that that this `.unwrap()` is totally safe but I think it's good to add a very short comment saying why. (Should make it easier in future to determine which `.unwrap()`'s we should try to remove and which we can ignore).
```suggestion
                .unwrap() // This won't panic as `self.log_size() - log_chunks < self.log_size() < Val::TWO_ADICITY`
```

SyxtonPrime commented on commit/src/domain.rs:
```diff
@@ -283,12 +265,12 @@ impl<Val: TwoAdicField> PolynomialSpace for TwoAdicMultiplicativeCoset<Val> {
         LagrangeSelectors {
             is_first_row: single_point_selector(0),
-            is_last_row: single_point_selector((1 << self.log_n) - 1),
+            is_last_row: single_point_selector((1 << self.log_size()) - 1),
```
```suggestion
            is_last_row: single_point_selector(self.size() - 1),
```

Antonio95 replied:
The proposed code complained that `single_point_selector` expects a `u64` but `self.size()` returns a `usize`. We have added a hard cast (cf https://github.com/Plonky3/Plonky3/blob/31cfc8b8be83c16500ffbf9f65515ec6b2b83041/commit/src/domain.rs#L274).

This is functionally the same as the original code `single_point_selector((1 << self.log_size()) - 1)` in the sense that both panic in the exact same situations. Let us know if the final code looks good to you or you would prefer a different solution and/or a comment.

SyxtonPrime replied:
Hmm yeah I think this looks good. There is, in principle, a small difference between the option if `usize` is not equivalent to a `u64` but I don't think that should matter. None of our cosets should be bigger than `2^31` anyway.

SyxtonPrime commented on commit/src/domain.rs:
```diff
@@ -283,12 +265,12 @@ impl<Val: TwoAdicField> PolynomialSpace for TwoAdicMultiplicativeCoset<Val> {
                 .into_iter()
                 .cycle()
-                .take(1 << coset.log_n)
+                .take(1 << coset.log_size())
```
```suggestion
                .take(coset.size())
```

SyxtonPrime reviewed (approved):
This all looks great, thank you for all your work on this.

Basically all my remaining comments are tiny nits/typo fixes so I think we are good to merge this once the CI issue is sorted out.

Cesar199999 commented:
> Ahh I think this might be coming from a clash with PR #740 which slightly changed how RNG worked. I think this should be an easy fix, just add the imports:
> 
> ```
> use rand::rngs::SmallRng;
> use rand::{Rng, SeedableRng};
> ```
> 
> and then explicitly initialize the rng with
> 
> ```
> let mut rng = SmallRng::seed_from_u64(1);
> ```
> 
> (Feel free to use a seed other than `1` if you want).
> 
> (I suspect the reason its working on your devices it you haven't merged in the recent main changes. The CI pipeline is run on the result of merging the PR into main so sometimes when a PR is slightly behind main there are clashing changes that cause these sorts of odd problems).

That makes a lot of sense! Our main misconception was thinking the CI pipeline ran directly on the PR code, rather than on the merge of the PR code with the main branch. Your comment was a very helpful clarification.

We've implemented the changes with the proper imports and explicit RNG initialization (we went ahead with many different seeds), and it's fixed now. Thanks for pointing us in the right direction.

Antonio95 commented:
Thank you @SyxtonPrime for the very thorough reviewing. We really think the code has improved thanks to it!
